### ISSUE TITLE
expose ratified gov actions in ledger state

### ISSUE THREAD

**Issue by ashisherc** [IntersectMBO/cardano-ledger#4491]
in ledger state we have enacted gov actions and expired gov actions, we would also like to have ratified gov actions

**Comment by Ryun1** [comment]
To be clear, I believe the ask here is for the ledger state to return the governance action IDs of actions within the `nextRatifyState`, rather than just the next whole next state.

**Comment by ashisherc** [comment]
@Ryun1 thats correct!

**Comment by kevinhammond** [comment]
Could you provide some more detail for the team, please, so they can evaluate the effort involved

- Exactly what information would you like - we assume governance action ids for actions that have been ratified but which are not yet enacted?

Is that correct?

Thanks!

**Comment by disassembler** [comment]
from the user perspective ratified and enacted are the same. They're separate rules in the ledger for technical reasons but happen at the same time.

e.g. that means you can never have an action that's ratified but not enacted.

Can you restate what it is that the enacted state doesn't give you?

**Comment by lehins** [comment]
Sorry @ashisherc, but we do need a bit more information.

First of all there isn't really such thing in ledger as ratified governance actions, because ratified and enacted are one and the same thing. An action can't be ratified without being enacted.

Second of all ledger is a library so it doesn't return anything and the whole of ledger state is exposed, so nothing is really hidden. Maybe you are trying to say that you need a ledger state query  that would expose some part of the ledger state that cardano-node should implement?

> return the governance action IDs of actions within the nextRatifyState, rather than just the next whole next state.

Here is the `RatifyState`, which part of it do you need?
https://github.com/IntersectMBO/cardano-ledger/blob/0274cf65dbb79773122b69dfd36a8299eec2783f/eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Internal.hs#L262-L268

**Comment by ashisherc** [comment]
referring to the section, `Gov Action Lifecycle` and `Changes to the local state-query protocol`, I assumed that the actions are ratified at an epoch boundary, visible throughout the epoch and then enacted at the next epoch boundary., being visible for 1 epoch duration. According to the explanation above, That's my misunderstanding and explains that the ratification is done just before the epoch boundary and enacted after the epoch boundary, so users don't see the ratification state.

This clears everything. Thanks much for looking at this and your time. 🙏

**Comment by ashisherc** [comment]
@lehins if the actions are ratified/enacted almost at the same time (during the epoch boundary), what does the `nextEnactState` in ledger state refer to?

also what are the `enactedGovActions` in the ledger state, enacted in the current epoch (applied to ledger state)?

**Comment by ashisherc** [comment]
@lehins also `nextEnactState` always has a state(data) even if the `enactedGovActions` is empty, what does nextEnactState refer to?

**Comment by lehins** [comment]
Unfortunately I can't answer theses questions simply without explaining how this process actually works. So, here is the outline of how proposal enactment is implemented:

1. Governance actions are proposed
2. Through potentially many epochs all the relevant entities vote on those proposals
3. On every epoch boundary we take a snapshot of the world, so it can no longer be affected by any of the transactions throughout the next epoch. Let's look at some particular example of us taking a snapshot at the very beginning of the epoch E(n), at which point ledger state contains three proposals: P(0), P(1) and P(2). All of them might have some votes associated, but only P(1) has enough votes to be enacted, while P(2) does not have enough and is set to expire in epoch E(n). 
4. At this point it is important to note that we already have all the information necessary to make a decision which actions will be ratified and enacted, as well as which ones will be removed from the state due to expiry. However, there are multiple reasons why we can't enact those actions at this epoch boundary E(n). One of those reasons is that this computation is expensive, so we spread it out throughout the epoch.
5. Somewhere in the first half of the epoch E(n) (in particular two stability windows before the end of the epoch) we are done with counting up the stake distribution. At that point we execute RATIFY and ENACT rules and record the outcome of that computation in the `nextEnactState`. This means that at this point we record the fact that P(2) will be removed due to expiry and P(1) will be enacted, as well as record the affect of applying P(1) in the `EnactState`. In other words `nextEnactState` has all of enacted actions applied to it, but it is still not yet in affect.
6. That is the step when the enacted governance actions go into affect. At the very beginning of epoch E(n+1) we apply the `nextEpochState` from step 5. to the ledger state, remove all of the enacted an expired proposals, refund the deposits, etc. and go back to the step 1.


> if the actions are ratified/enacted almost at the same time (during the epoch boundary), what does the nextEnactState in ledger state refer to?

This is the state that will be applied at the next epoch boundary.

> also what are the `enactedGovActions` in the ledger state, enacted in the current epoch (applied to ledger state)?

These are the actions that where ratified and enacted, i.e. applied to the `nextEnactState`, which itself will be applied at the epoch boundary.

> `nextEnactState` always has a state(data) even if the `enactedGovActions` is empty, what does nextEnactState refer to?

If `enactedGovActions` is empty, then `nextEnactState` will have no changes when compared to the previous epoch, i.e. nothing was enacted.

So, to sum it up. Ratification and enactment is a multistage process with the final stage being the application of `nextEnactState` to the ledger state at the epoch boundary.

**Comment by ashisherc** [comment]
@lehins thanks much for the detailed explanation, that clears everything. 🙏

**Comment by ashisherc** [comment]
@lehins does the ledger allow submitting votes on an action after it has been enacted(applied), or has expired?

**Comment by ashisherc** [comment]
@lehins 

> On every epoch boundary we take a snapshot of the world, so it can no longer be affected by any of the transactions throughout the next epoch

Which snapshot is used for votes calculation? Assuming from this message it should be the Mark, but it does not maintain stakeKey->dRep delegation.

**Comment by ashisherc** [comment]
@lehins it will be great if you can also detail how votes are computed for ratification. Eg. votes submitted before the epoch boundary (before the epoch in which pulser will ratify actions), is so, is this available in the ledger state?

**Comment by lehins** [comment]
> does the ledger allow submitting votes on an action after it has been enacted(applied), or has expired?

No.

> Which snapshot is used for votes calculation? Assuming from this message it should be the Mark, but it does not maintain stakeKey->dRep delegation.

Mark, Set, Go snapshots are unrelated to the snapshot I was talking about. The one that I was describing is literally snapshot of the whole world relevant to governance.

> it will be great if you can also detail how votes are computed for ratification.

There is too much information to describe here. You can find it by following these functions: https://github.com/IntersectMBO/cardano-ledger/blob/690c128ea48204894a80731265ef782cdfb6a0d2/eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs#L315-L317

>  is this available in the ledger state?

There is no ledger state query, but it is available in the state. Here is how you can access it programmatically (`psProposals` is the field where all of the votes are stored): https://github.com/IntersectMBO/cardano-ledger/blob/master/libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/Query.hs#L312-L316

**Comment by ashisherc** [comment]
@lehins 
> psProposals is the field where all of the votes are stored

It stores drep id yes/no, stake pool yes/no

but how do you compute the stake delegated to dRep, spo during the ratification (As am unable to find it in the ledger state). I believe the votes ratification requires ADA delegated amount.

--

where can I find the `reDRepState`, https://github.com/IntersectMBO/cardano-ledger/blob/690c128ea48204894a80731265ef782cdfb6a0d2/eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ratify.hs#L237 from ledger state?

similarly for stake pool as well, if you suggest Mark is not used for this purpose.

**Comment by lehins** [comment]
> similarly for stake pool as well, if you suggest Mark is not used for this purpose.

Stake distribution from "Mark" is being reused for counting up the votes for stake pools, however that is just a small part of the snapshot that is taken for governance:

https://github.com/IntersectMBO/cardano-ledger/blob/1af4445646a9a14594b3ced917773d568a7334d8/eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs#L309

This is where the actual stake distribution for votes is counted up: https://github.com/IntersectMBO/cardano-ledger/blob/1af4445646a9a14594b3ced917773d568a7334d8/eras/conway/impl/src/Cardano/Ledger/Conway/Governance/DRepPulser.hs#L297

https://github.com/IntersectMBO/cardano-ledger/blob/1af4445646a9a14594b3ced917773d568a7334d8/eras/conway/impl/src/Cardano/Ledger/Conway/Governance/DRepPulser.hs#L208-L243

**Comment by ashisherc** [comment]
@lehins are the dRep related distr available in ledger state dump, as used in the calculations above?

**Comment by ashisherc** [comment]
@lehins ping!

**Comment by lehins** [comment]
> @lehins ping!

@ashisherc pong!

> are the dRep related distr available in ledger state dump, as used in the calculations above?

Yes: https://github.com/IntersectMBO/cardano-ledger/blob/3c1b5765cc219e85988bafa8d5e546e45ae9c1a1/libs/cardano-ledger-api/src/Cardano/Ledger/Api/State/Query.hs#L151-L164

**Comment by ashisherc** [comment]
@lehins thanks for the clarification of this query. this distribution is not available in the ledger-state query (full dump), is it intentional?

**Comment by lehins** [comment]
> this distribution is not available in the ledger-state query (full dump), is it intentional?

Full LedgerState dump is intended for testing and development only. So, whatever is in there is intentional.

**Comment by ashisherc** [comment]
@lehins so am saying dRep distr is NOT available in the full dump,

**Comment by ashisherc** [comment]
@lehins if a snapshot is taken for the governance world, are the votes parts of it? if so can we query them?

if not, do votes made during the ratification process are considered if they are submitted just before pulser is executing ratification?
